bpo-45512: Simplify isolation_level handling in `sqlite3`

erlend-aasland opened the pull request:
Simplify reference handling and merge two connection object members:
- Remove superfluous `PyObject *` member `isolation_level` from connection object
  (self->begin_statement carries the information we need)
- Introduce `get_begin_statement()` helper for converting isolation level to begin statement
- Simplify `sqlite3.Connection.__init__` by using `get_begin_statement`
- Use AC to remove reference handling for the isolation level parameter

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- issue-number: [bpo-45512](https://bugs.python.org/issue45512) -->
https://bugs.python.org/issue45512
<!-- /issue-number -->

Commit:
Remove connection isolation_level

Commit:
Factor out get_begin_statement()

Commit:
Use get_begin_statment() in connection __init__

Commit:
Clean up

Commit:
Merge branch 'main' into sqlite-isolation-level

Regenerate clinic

Commit:
Merge branch 'main' into sqlite-isolation-level

Commit:
Merge branch 'main' into sqlite-isolation-level

Commit:
Simplify exception messages; get rid of ref leak

Commit:
Make exception string more compact

corona10 commented on Modules/_sqlite/connection.c:
```diff
@@ -107,6 +106,21 @@ new_statement_cache(pysqlite_Connection *self, int maxsize)
+    assert(level != NULL);
+    for (int i = 0; begin_statements[i] != NULL; i++) {
+        const char *candidate = begin_statements[i] + 6;
+        if (strlen(candidate) == (size_t)sz) {
```
Do we really need a length check here?
I mean that `sqlite3_stricmp()` will crash without it?

erlend-aasland replied:
Yes, unfortunately. There are some corner cases with embedded null characters that will fail. The test suite already got regression tests for these.

Note: it will not crash; it will just let through some false positives.

erlend-aasland replied:
What I could do is to check for embedded null characters in `pysqlite_connection_set_isolation_level`. Then we can remove the strlen check (and the size argument). Might be a better solution.

Note: when `get_begin_statement` is called from `__init__`, the embedded null char check has already been done by AC, so we _know_ we're safe in that path.

erlend-aasland replied:
I pushed this change; I think it makes for more readable code:

- `get_begin_statement()` is clearer
- there is no need to call `strlen()` in `__init__`

Commit:
Merge branch 'main' into sqlite-isolation-level

Commit:
Check for embedded nulls in isolation_level setter

This simplifies get_begin_statement(), and removes the need to call
strlen() in __init__()

Commit:
Merge branch 'main' into sqlite-isolation-level

Regenerate clinic

Commit:
Add helper to improve understanding between begin statement and isolation level
